sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala

```diff
@@ -39,17 +39,45 @@ import scala.annotation.tailrec
   *   [[ResultNeedPackage]] continuation is called.
   * @param forbidLocalContractIds when `true` the preprocessor will reject
   *   any value/command/transaction that contains a local Contract ID.
+  * @param costModel the preprocessor input cost model that should be used
+  * @param initialInputCost initial input cost
+  * @param maxInputCost maximum input cost. Any input cost that exceeds this value will cause engine evaluation to
+  *   terminate with an error.
   */
 private[engine] final class Preprocessor(
     compiledPackages: CompiledPackages,
     loadPackage: (Ref.PackageId, language.Reference) => Result[Unit],
     forbidLocalContractIds: Boolean = true,
+    costModel: CostModel.CostModelImplicits = CostModel.EmptyCostModelImplicits,
+    initialInputCost: CostModel.Cost = 0L,
+    maxInputCost: CostModel.Cost = Long.MaxValue,
```

Comment:
Intentionally opted for the preprocessor being configured with a `maxInputCost` so that preprocessing could be terminated before command evaluation could start.